src/s3/commands/uploadFile.ts

```diff
@@ -114,11 +142,207 @@ async function uploadWithProgress({
     )
 }
 
-function statFile(file: vscode.Uri) {
-    return statSync(file.fsPath).size
+interface BucketQuickPickItem extends vscode.QuickPickItem {
+    bucket: S3.Bucket | undefined
 }
 
-async function refreshNode(node: S3BucketNode | S3FolderNode, commands: Commands): Promise<void> {
-    node.clearChildren()
-    return commands.execute('aws.refreshAwsExplorerNode', node)
+/**
+ * Will display a quick pick with the list of all buckets owned by the user.
+ * 
+ * @returns Bucket selected by the user, undefined if no bucket was selected or the quick pick was cancelled.
+ * 
+ * @throws Error if there is an error calling s3
+ */
+export async function promptUserForBucket(
+    s3client: S3Client,
+    window = Window.vscode(),
+    promptUserFunction = promptUser,
+    createBucket = createBucketCommand,
+): Promise<S3.Bucket | string> {
+    let allBuckets: S3.Bucket[]
+    try{
+        allBuckets = await s3client.listAllBuckets()
+    } catch (e) {
+        getLogger().error('Failed to list buckets from client', e)
+        showErrorWithLogs('Failed to list buckets from client', window)
+        telemetry.recordS3UploadObject({ result: 'Failed' })
+        throw new Error('Failed to list buckets from client')
+    }
+    
+    const s3Buckets = allBuckets.filter(bucket => {
+        return bucket && bucket.Name
+    }) as S3.Bucket[]
+
+    
+    const createNewBucket: BucketQuickPickItem = {
+        label: localize('AWS.command.s3.createBucket',"Create new bucket"),
+        bucket: undefined
+    }
+    const bucketItems: BucketQuickPickItem[] = s3Buckets.map(bucket => {
+        return { 
+            label: bucket.Name!,
+            bucket
+        }
+    })
+
+    const picker = createQuickPick({
+        options: {
+            canPickMany: false,
+            ignoreFocusOut: true,
+            title: localize('AWS.message.selectBucket','Select an S3 bucket to upload to'),
+            step: 2,
+            totalSteps: 2,
+        },
+        items: [...bucketItems, createNewBucket],
+        buttons: [vscode.QuickInputButtons.Back]
+    })
+    const response = verifySinglePickerOutput(await promptUserFunction({
+        picker: picker,
+        onDidTriggerButton: (button, resolve, reject) => {
+            if (button === vscode.QuickInputButtons.Back) {
+                resolve([{
+                    label: "back",
+                    bucket: undefined
+                }])
+            }
+
+        },
+    })) 
+
+    if (!response) {
+        return 'cancel'
+    }
+
+    if (!response.bucket) {
+        if (response.label === 'back') {
+            return response.label
+        }
+        if (response.label === 'Create new bucket') {
+            const s3Node = new S3Node(s3client)
+            await createBucket(s3Node, window, Commands.vscode())
+            return promptUserForBucket(s3client)
+        }
+    } else {
+        return response.bucket
+    }
+    return 'cancel'
```

Comment:
I don't think this line will ever be hit.

Reply:
That is true

Reply:
Could you remove this line?

Reply:
It actually gives an error if i remove it

Reply:
Compiler probably doesn't recognize that you're checking for undefined in your if statements.